### PR TITLE
Address space changes

### PR DESCRIPTION
**Pull request by rbradford**
This PR addresses some of the issues that we have with our address space handling. The main limitation it addresses is safely dividing the space between RAM and devices, it does this by correctly patching the ACPI tables so that the address space range correctly represents the device range (top_of_address_space - top_of_ram). This was already working correctly when not using ACPI as the E820 tables for RAM ended at this point.

The other major change in this PR is to increase the address space based on the size of the host physical address space. This rectifies limitations in how much RAM can be given to VMs or used with virtio-pmem.

There are also minor fixes to the E820 and ACPI tables that were observed during this development.

Fixes: #286

### CODE REVIEW

File: arch/src/x86_64/mod.rs [hunk 1]
```diff
@@ -130,6 +130,7 @@ pub fn configure_system(
     num_cpus: u8,
     setup_hdr: Option<setup_header>,
     _serial_enabled: bool,
+    _end_of_range: GuestAddress,
```
Comment (sboeuf) [inline arch/src/x86_64/mod.rs]: That's actually nice to provide the end of range, and I would suggest that we use it to replace every occurrence of `guest_mem.end_addr()`. This will remove the assumption that `GuestMemoryMmap` is only about RAM.

File: arch/src/x86_64/mod.rs [hunk 2]
```diff
@@ -166,7 +166,7 @@ pub fn configure_system(
         add_e820_entry(
             &mut params.0,
             himem_start.raw_value(),
-            mem_end.unchecked_offset_from(himem_start),
+            mem_end.unchecked_offset_from(himem_start) + 1,
```
Comment (sboeuf) [inline arch/src/x86_64/mod.rs]: Yes the last byte was missing, and you could say in the commit message that's because we're providing the range size here.

File: vm-virtio/src/transport/pci_device.rs [hunk 3]
```diff
@@ -239,6 +239,9 @@ pub struct VirtioPciDevice {
 
     // Setting PCI BAR
     settings_bar: u8,
+
```
Comment (sboeuf) [inline vm-virtio/src/transport/pci_device.rs]: Just some general questions about your commit message here. You said:
```
Currently the hypervisor sets up an identify mapping of 4GiB which the firmware extends to 64GiB
```
- Why is the firmware actually extending from 4G to 64G?
- Why is this extension needed?
- Why 64G and not 32G or 128G for instance?

File: vmm/src/vm.rs [hunk 4]
```diff
@@ -676,7 +676,7 @@ impl<'a> Vm<'a> {
             GuestAddress(0),
             1 << 16 as GuestUsize,
             GuestAddress(0),
-            1 << 36 as GuestUsize,
+            1 << 36,
```
Comment (sameo) [inline vmm/src/vm.rs]: Nit: This change is unrelated to that patch imho.

File: arch/src/x86_64/mod.rs (hunk 1 above)
Comment (rbradford) [reply arch/src/x86_64/mod.rs]: But those two are not the the same. `end_of_range` is the last byte in the entire address space.  `guest_mem.end_addr()` is the size of the RAM minus 1. Or it is at least at the point it is used to calculate the E820 tables in `configure_system()`. 

Your point is still valid though, we should look at having a different, explicit parameter to store the `end_of_ram` in particular if we start to use `GuestMemoryMmap` for handing the currently manual mmap()s then it will go wrong.

File: vm-virtio/src/transport/pci_device.rs (hunk 3 above)
Comment (rbradford) [reply vm-virtio/src/transport/pci_device.rs]: The hypervisor only creates an identity mapping for 1GiB (not 4GiB, sorry my mistake). You need the extension as you always needs a identity page table entries to access any memory. Previously with our 64GiB address space limit that would require extending to 64GiB to be able to access the PCI devices as they were put at the top of that address space.

File: arch/src/x86_64/mod.rs (hunk 1 above)
Comment (rbradford) [reply arch/src/x86_64/mod.rs]: I opened #289 to track this.

Comment (sboeuf) [reply arch/src/x86_64/mod.rs]: Great thanks!

File: vm-virtio/src/transport/pci_device.rs (hunk 3 above)
Comment (sboeuf) [reply vm-virtio/src/transport/pci_device.rs]: >Previously with our 64GiB address space limit that would require extending to 64GiB to be able to access the PCI devices as they were put at the top of that address space.

Yes so if we end up with an address space of 1TiB, we need the extension to 1TiB now, right? I mean your patch didn't change the behavior about where to place the devices (they're still at the end of the address space), except for the virtio-blk one.

Comment (rbradford) [reply vm-virtio/src/transport/pci_device.rs]: We don't need it because the block devices are in the 32-bit gap (this was the whole point of moving the block devices.)

Comment (sboeuf) [reply vm-virtio/src/transport/pci_device.rs]: Ah okay now I understand what you meant. We were extending to 64GiB, but we would have needed to extend to even bigger when we started to look at supporting real physical address space, and instead we simplified the whole problem by moving block and not extending at all in the firmware.
